### ISSUE TITLE
Wrong `if` in `bindsnet.network.topology.LocalConnection.compute`

### ISSUE THREAD

**Issue by danielgafni** [BindsNET/bindsnet#435]
Hello!

I'm working with a LocalConnection.
Please, check its `compute` method... As I understand, the first part of the `if` is wrong. The tensor it returns has a wrong shape for target layers of shapes with more than 2 dimensions. I'm providing a minimal code (I'm running it from a Jupyter Notebook) example that breaks for me:

```python
from torchvision.datasets import FashionMNIST
fashion_mnist = FashionMNIST(root="fasion-mnist", download=True)

from bindsnet.encoding import PoissonEncoder
import torch
from torchvision import transforms

time_max = 250
dt = 1
intensity = 127.5
crop = 20

# DataLoader
class EncodedFashionMnist(torch.utils.data.Dataset):
    def __init__(self, data, crop=20, time=250, intensity=127.5, dt=1):
        self.data = data
        self.crop = crop
        self.time = time
        self.intensity = intensity
        self.dt = dt
        
        self.center_crop = transforms.CenterCrop(20)
        self.poisson_encoder = PoissonEncoder(time=time_max, intensity=intensity, dt=dt)
        self.transform = transforms.Compose([
            self.center_crop, transforms.ToTensor(), self.poisson_encoder
        ])
        
        
    def __getitem__(self, i):
        return {"encoded_image": self.transform(self.data[i][0]), "label": self.data[i][1]}
    
    def __len__(self):
        return len(self.data)

encoded_fashion_mnist = EncodedFashionMnist(fashion_mnist)

train_dataloader = torch.utils.data.DataLoader(
    encoded_fashion_mnist,
    batch_size=1,
)

# Network
from bindsnet.network import Network
from bindsnet.learning import PostPre
from bindsnet.network.monitors import Monitor
from bindsnet.network.nodes import AdaptiveLIFNodes, Input
from bindsnet.network.topology import LocalConnection, Connection
from bindsnet.analysis.plotting import (
    plot_input,
    plot_spikes,
    plot_conv2d_weights,
    plot_voltages,
)
import numpy as np

# Hyperparameters
padding = 0
kernel_size = 12
stride = 4
conv_size = int((crop - kernel_size + 2 * padding) / stride) + 1
tc_decay = 20.0
thresh = -52
refrac = 2
wmin = 0
wmax = 1
mean_weight = 0.4
n_filters = 25
tau_pos = 12.
tau_neg = 5.

network = Network(learning=True)
n_input = crop ** 2
input_layer = Input(
    n=n_input, shape=(1, crop, crop), traces=True, refrac=refrac
)
n_output = n_filters * conv_size * conv_size
output_shape = int(np.sqrt(n_output))
output_layer = AdaptiveLIFNodes(
    n=n_output,
    shape=(n_filters, conv_size, conv_size),
    traces=True,
    thres=thresh,
    tc_trace_pre=tau_pos,
    tc_trace_post=tau_neg,
    tc_decay=tc_decay,
    theta_plus=0.05,
    tc_theta_decay=1e6,
    refrac=refrac,
)

kernel_prod = kernel_size ** 2

norm = mean_weight * kernel_prod

connection_XY = LocalConnection(
    input_layer,
    output_layer,
    n_filters=n_filters,
    kernel_size=kernel_size,
    stride=stride,
    update_rule=PostPre,
    norm=norm,
    nu=[1e-4, 1e-2],
    wmin=wmin,
    wmax=wmax,
)

network.add_layer(input_layer, name="X")
network.add_layer(output_layer, name="Y")
network.add_connection(connection_XY, source="X", target="Y")

from tqdm.notebook import tqdm

for batch in tqdm(train_dataloader):
    network.run(inputs={"X": batch["encoded_image"].squeeze(0)}, time=time_max, input_time_dim=0)
```

So here is the method I'm talking about:

```python
    def compute(self, s: torch.Tensor) -> torch.Tensor:
        # language=rst
        """
        Compute pre-activations given spikes using layer weights.
        :param s: Incoming spikes.
        :return: Incoming spikes multiplied by synaptic weights (with or without
            decaying spike activation).
        """
        # Compute multiplication of pre-activations by connection weights.
        if self.w.shape[0] == self.source.n and self.w.shape[1] == self.target.n:
            return s.float().view(s.size(0), -1) @ self.w + self.b
        else:
            a_post = (
                s.float().view(s.size(0), -1)
                @ self.w.view(self.source.n, self.target.n)
                + self.b
            )
            return a_post.view(*self.target.shape)
```

My code example above works if I change it to:

```python
    def compute(self, s: torch.Tensor) -> torch.Tensor:
        # language=rst
        """
        Compute pre-activations given spikes using layer weights.

        :param s: Incoming spikes.
        :return: Incoming spikes multiplied by synaptic weights (with or without
            decaying spike activation).
        """
        # Compute multiplication of pre-activations by connection weights.
        a_post = (
            s.float().view(s.size(0), -1)
            @ self.w.view(self.source.n, self.target.n)
            + self.b
        )
        return a_post.view(*self.target.shape)
```

So basically the first part of the `if` is wrong. This check is not working for layer shapes of more than two dimensions, probably. In my case:

the output must have shape: (25, 3, 3)
but it has shape: (1, 225)

I have:
source.n: 400
target.n: 225
self.w shape: (400, 225)

So the check in this `if` is taking this case wrong, because actually my Y layer shape is complex, not just (1, 225). 

The second part of the if takes care of it by doing `.view(*self.target.shape)`. 
Maybe, this should be the default behavior? I can submit a pull request.
Please tell me if I need to clarify something or provide any files or logs.

**Comment by Hananel-Hazan** [comment]
Thanks @danielgafni for finding this.
Please submit a PR with the corrections.

**Comment by danielgafni** [comment]
@Hananel-Hazan my PR is #436. Thanks!

**Comment by Hananel-Hazan** [comment]
@danielgafni Thanks for the PR!
